client/eth/connection_pool.go

```diff
@@ -99,21 +122,27 @@ func (c *ConnectionPoolImpl) DialContext(ctx context.Context, _ string) error {
 }
 
 func (c *ConnectionPoolImpl) GetHTTP() (Client, bool) {
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
-retry:
-	_, client, ok := c.cache.GetOldest()
-	if !client.Health() {
-		goto retry
+	if c.cache == nil {
+		return nil, false
 	}
-	return client, ok
+	return c.getClient(c.cache)
 }
 
 func (c *ConnectionPoolImpl) GetWS() (Client, bool) {
+	// Because the WS URL is optional, we need to check if it's nil.
+	if c.wsCache == nil {
+		return nil, false
+	}
+	return c.getClient(c.wsCache)
+}
+
+func (c *ConnectionPoolImpl) getClient(
```

Comment:
```suggestion
func (c *ConnectionPoolImpl) getClientFrom(
```
to indicate that this helper function is using the provided `cache` LRU

Reply:
also the locking should not happen here. Add a comment above this function something like "NOTE: this function assumes the lock is held"